R/compute_metrics.R

```diff
@@ -0,0 +1,160 @@
+#' have been supplied when generating `x`.
+#'
+#' If the `metrics` argument is equivalent to that passed as the `metrics`
+#' argument when generating `x`, then `compute_metrics(x, ...)` will return
+#' the same results as `collect_metrics(x, ...)` (though will be quite a
+#' bit slower).
+#'
+#' @param x The results of a tuning function like [tune_grid()] or
+#' [fit_resamples()], generated with the control option `save_pred = TRUE`.
+#' @param metrics A [metric set][yardstick::metric_set()] of new metrics
+#' to compute. Each metric in the set must have a metric type (usually
+#' `"numeric"`, `"class"`, or `"prob`) that matches some metric evaluated
+#' when generating `x`.
+#' @param summarize A single logical value indicating whether metrics should
+#' be summarized over resamples (`TRUE`) or return the values for each
+#' individual resample. See [collect_metrics()] for more details on how
+#' metrics are summarized.
+#' @inheritParams control_grid
+#' @inheritParams collect_metrics
+#'
+#' @return A tibble. See [collect_metrics()] for more details on the return value.
+#'
+#' @examplesIf tune:::should_run_examples()
+#' # load needed packages:
+#' library(parsnip)
+#' library(rsample)
+#' library(yardstick)
+#'
+#' # evaluate a linear regression against resamples.
+#' # note that we pass `save_pred = TRUE`:
+#' res <-
+#'   fit_resamples(
+#'     linear_reg(),
+#'     mpg ~ cyl + hp,
+#'     bootstraps(mtcars, 5),
+#'     control = control_grid(save_pred = TRUE)
+#'   )
+#'
+#' # to return the metrics supplied to `fit_resamples()`:
+#' collect_metrics(res)
+#'
+#' # to compute new metrics:
+#' compute_metrics(res, metric_set(mae))
+#'
+#' # if `metrics` is the same as that passed to `fit_resamples()`,
+#' # then `collect_metrics()` and `compute_metrics()` give the same
+#' # output, though `compute_metrics()` is quite a bit slower:
+#' all.equal(
+#'   collect_metrics(res),
+#'   compute_metrics(res, metric_set(rmse, rsq))
+#' )
+#' @name compute_metrics
+#' @export
+compute_metrics <- function(x, metrics, summarize, event_level, ...) {
+  UseMethod("compute_metrics")
+}
+
+#' @export
+#' @rdname compute_metrics
+compute_metrics.default <- function(x, metrics, summarize, event_level, ...) {
+  rlang::abort("No `compute_metrics()` method exists for this type of object.")
+}
+
+#' @export
+#' @rdname compute_metrics
+compute_metrics.tune_results <- function(x,
+                                         metrics,
+                                         summarize = TRUE,
+                                         event_level = "first",
+                                         ...) {
+  if (!any(names(x) == ".predictions")) {
+    rlang::abort(paste0(
+      "The inputted `x` argument must have been generated with the ",
+      "control argument `save_pred = TRUE`."
+    ))
+  }
+
+  if (!inherits(metrics, "metric_set")) {
+    rlang::abort("The inputted `metrics` argument must be a metric set.")
```

Comment:
```suggestion
    rlang::abort("`metrics` must be a metric set.")
```

Would be a good place to use the non-existent `check_inherits(metrics, "metric_set")` type checker. Just pointing it out for myself